storages/multipart_round_tripper.go

```diff
@@ -22,91 +23,89 @@ import (
+func (multiPartRoundTripper MultiPartRoundTripper) Cancel() error { return nil }
 
+// newMultiPartRoundTripper initializes multipart client
+func newMultiPartRoundTripper(backends []*Backend) client {
+	multiPartRoundTripper := &MultiPartRoundTripper{}
 	var backendsEndpoints []string
 	var activeBackendsEndpoints []string
 
 	multiPartRoundTripper.backendsRoundTrippers = make(map[string]*Backend)
 
-	for _, roundTripper := range backends {
-
-		if backend, isBackendType := roundTripper.(*Backend); isBackendType {
-
-			if !backend.Maintenance {
-				multiPartRoundTripper.backendsRoundTrippers[backend.Endpoint.Host] = backend
-				activeBackendsEndpoints = append(activeBackendsEndpoints, backend.Endpoint.Host)
-			}
-
-			backendsEndpoints = append(backendsEndpoints, backend.Endpoint.Host)
+	for _, backend := range backends {
+		if !backend.Maintenance {
+			multiPartRoundTripper.backendsRoundTrippers[backend.Endpoint.Host] = backend
+			activeBackendsEndpoints = append(activeBackendsEndpoints, backend.Endpoint.Host)
 		}
+
+		backendsEndpoints = append(backendsEndpoints, backend.Endpoint.Host)
 	}
 
 	multiPartRoundTripper.backendsEndpoints = backendsEndpoints
 	multiPartRoundTripper.backendsRing = hashring.New(activeBackendsEndpoints)
+	return multiPartRoundTripper
 }
 
-// RoundTrip performs a RoundTrip using the strategy described in MultiPartRoundTripper
-func (multiPartRoundTripper *MultiPartRoundTripper) RoundTrip(request *http.Request) (response *http.Response, requestError error) {
-
-	if isMultiPartUploadRequest(request) {
-
-		if !multiPartRoundTripper.canHandleMultiUpload() {
-			log.Debugf("Multi upload for %s failed - no backends available.", request.URL.Path)
-			return nil, errors.New("Can't handle multi upload")
-		}
-
-		multiUploadBackend, backendSelectError := multiPartRoundTripper.pickBackend(request.URL.Path)
-
-		if backendSelectError != nil {
-			log.Debugf("Multi upload failed for %s - %s", backendSelectError, request.URL.Path)
-			return nil, errors.New("Can't handle multi upload")
-		}
+// Do performs backend request
+func (multiPartRoundTripper *MultiPartRoundTripper) Do(request *http.Request) <-chan BackendResponse {
+	out := make(chan BackendResponse)
+	if !multiPartRoundTripper.canHandleMultiUpload() {
+		log.Debugf("Multi upload for %s failed - no backends available.", request.URL.Path)
+		go func() {
+			out <- BackendResponse{Response: nil, Error: errors.New("Can't handle multi upload")}
+			close(out)
+		}()
+		return out
+	}
 
-		log.Debugf("Handling multipart upload, sending %s to %s, RequestID id %s",
-			request.URL.Path,
-			multiUploadBackend.Endpoint,
-			request.Context().Value(log.ContextreqIDKey))
+	multiUploadBackend, backendSelectError := multiPartRoundTripper.pickBackend(request.URL.Path)
 
-		response, requestError = multiUploadBackend.RoundTrip(request)
+	if backendSelectError != nil {
+		log.Debugf("Multi upload failed for %s - %s", backendSelectError, request.URL.Path)
+		go func() {
```

Comment:
As above - why we need a separate go-routine?

Reply:
As above: yes